src/ast/ast.h

```diff
@@ -272,25 +273,27 @@ class Integer : public Node {
   explicit Integer(ASTContext &ctx,
                    Location &&loc,
                    uint64_t n,
-                   bool force_unsigned = false,
                    std::optional<std::string> &&original = std::nullopt)
       : Node(ctx, std::move(loc)),
-        integer_type(force_unsigned || n > std::numeric_limits<int64_t>::max()
-                         ? CreateUInt64()
-                         : CreateInt64()),
         value(n),
-        original(std::move(original)) {};
+        original(std::move(original)),
+        integer_type(util::get_integer_type(n)) {};
   explicit Integer(ASTContext &ctx, const Location &loc, const Integer &other)
       : Node(ctx, loc + other.loc),
-        integer_type(other.integer_type),
         value(other.value),
-        original(other.original) {};
+        original(other.original),
+        integer_type(other.type()) {};
 
   const SizedType &type() const
   {
     return integer_type;
   }
 
+  void setType(SizedType type) {
```

Comment:
Having this be mutable creates a lot more ambiguity. It seems like it is only used in two places: why don't you just have those two places inject a cast along with the literal integer instead? E.g. `ast.make_node<Cast>(CreateUint32(), ast.make_node<Integer>(value));`

Then you can drop `setType` and just rely on all the standard analysis and type coercion from the cast

Reply:
I didn't want to create extra codegen when we didn't need to. I had it before as a separate constructor for `Integer` and `NegativeInteger` - I'll just go back to that so it stays immutable.